### ISSUE TITLE
Rationalize/consolidate type parsers

### ISSUE THREAD

**Issue by jacques-n** [substrait-io/substrait#686]
There was originally a [type parser built in substrait-java](https://github.com/substrait-io/substrait-java/blob/main/core/src/main/antlr/SubstraitType.g4)

Later, someone took that and reimplemented it in [substrait-validator](https://github.com/substrait-io/substrait-validator/blob/main/rs/src/parse/extensions/simple/derivations/SubstraitType.g4)

We need to rationalize these two parsers (and any others that exist) and create a single parser in the core substrait repository.

I have no idea if they agree in behavior. I believe the java one also supports parsing to the other proto type definitions that aren't used in plans today (were intended for consumers informing producers of UDF behavior).

* [Parameterized types](https://github.com/substrait-io/substrait/blob/main/proto/substrait/parameterized_types.proto) (used in input arguments of extensions)
* [Type expressions](https://github.com/substrait-io/substrait/blob/main/proto/substrait/type_expressions.proto) (used in return type expressions of extensions)

**Comment by EpsilonPrime** [comment]
There is also two in substrait-cpp

**Comment by jacques-n** [comment]
![picard](https://cdn3.emoji.gg/emojis/facepalm.png)

Type parsers? Can you provide pointers to them.

**Comment by EpsilonPrime** [comment]
This has support for types and parameterized types:
https://github.com/substrait-io/substrait-cpp/tree/main/src/substrait/type

There's a wrapper with slight improvements for text parsing here:
https://github.com/substrait-io/substrait-cpp/blob/1dbf98b548de3ef11cac2a42075b87f57e7004b9/src/substrait/textplan/parser/SubstraitPlanTypeVisitor.cpp#L24

**Comment by jacques-n** [comment]
> Shouldn't we add the grammar definition to the core repo and keep the implementations in the language repos?

Yes, this was what this ticket is trying to suggest. I'll update description make clearer.

**Comment by nielspardon** [comment]
some good news on this: the grammar in substrait-java is gone and we're using the grammar from the substrait repo https://github.com/substrait-io/substrait-java/pull/730
